File: web/dashboard/src/lib/leakScanner.ts
import type { AuthSession } from "./auth";
import { serviceRequest } from "./serviceApi";

export type ScanTargetType = "git_repo" | "container_image" | "log_stream" | "s3_bucket" | "env_file";
export type FindingSeverity = "critical" | "high" | "medium" | "low" | "info";
export type FindingStatus = "open" | "acknowledged" | "resolved" | "false_positive";

export interface ScanTarget {
  id: string;
  name: string;
  type: ScanTargetType;
  uri: string;
  enabled: boolean;
  last_scanned_at?: string;
  created_at: string;
  scan_count: number;
  open_findings: number;
}

export interface ScanJob {
  id: string;
  target_id: string;
  target_name: string;
  target_type: ScanTargetType;
  status: "queued" | "running" | "completed" | "failed";
  started_at?: string;
  completed_at?: string;
  findings_count: number;
  error?: string;
  progress_pct?: number;
}

export interface LeakFinding {
  id: string;
  job_id: string;
  target_id: string;
  target_name: string;
  severity: FindingSeverity;
  type: string; // "aws_access_key", "generic_api_key", "jwt_token", etc.
  description: string;
  location: string; // file path, line number, etc.
  context_preview: string; // redacted snippet
  entropy: number;
  status: FindingStatus;
  detected_at: string;
  resolved_at?: string;
  resolved_by?: string;
  notes?: string;
}

export async function listTargets(session: AuthSession): Promise<ScanTarget[]> {
  const res = await serviceRequest<any>(session, "posture", "/leaks/targets");
  return res.items ?? [];
}

export async function createTarget(session: AuthSession, data: Partial<ScanTarget>): Promise<ScanTarget> {
  return serviceRequest<ScanTarget>(session, "posture", "/leaks/targets", { method: "POST", body: JSON.stringify(data) });
}

export async function deleteTarget(session: AuthSession, id: string): Promise<void> {
  return serviceRequest<void>(session, "posture", `/leaks/targets/${id}`, { method: "DELETE" });
}

export async function triggerScan(session: AuthSession, targetId: string): Promise<ScanJob> {
  return serviceRequest<ScanJob>(session, "posture", `/leaks/targets/${targetId}/scan`, { method: "POST" });
}

export async function listJobs(session: AuthSession): Promise<ScanJob[]> {
  const res = await serviceRequest<any>(session, "posture", "/leaks/jobs");
  return res.items ?? [];
}

export async function listFindings(session: AuthSession, params?: { status?: FindingStatus; severity?: FindingSeverity }): Promise<LeakFinding[]> {
  const q = new URLSearchParams();
  if (params?.status) q.set("status", params.status);
  if (params?.severity) q.set("severity", params.severity);
  const res = await serviceRequest<any>(session, "posture", `/leaks/findings?${q}`);
  return res.items ?? [];
}

export async function updateFinding(session: AuthSession, id: string, data: { status: FindingStatus; notes?: string }): Promise<LeakFinding> {
  return serviceRequest<LeakFinding>(session, "posture", `/leaks/findings/${id}`, { method: "PATCH", body: JSON.stringify(data) });
}

// Aliases for tab compatibility
export const listLeakTargets = listTargets;
export const addLeakTarget = createTarget;
export const deleteLeakTarget = deleteTarget;
export const triggerLeakScan = triggerScan;
export const listLeakFindings = listFindings;
export const listLeakJobs = listJobs;
export const resolveLeakFinding = updateFinding;

File: services/posture/leak_scanner.go
package main

import "time"

// LeakScanTarget is a registered source to be scanned for credential/secret leaks.
type LeakScanTarget struct {
	ID           string     `json:"id"`
	TenantID     string     `json:"tenant_id"`
	Name         string     `json:"name"`
	Type         string     `json:"type"` // git_repo, container_image, log_stream, s3_bucket, env_file
	URI          string     `json:"uri"`
	Enabled      bool       `json:"enabled"`
	LastScannedAt *time.Time `json:"last_scanned_at,omitempty"`
	CreatedAt    time.Time  `json:"created_at"`
	ScanCount    int        `json:"scan_count"`
	OpenFindings int        `json:"open_findings"`
}

// LeakScanJob represents an in-progress or completed scan job.
type LeakScanJob struct {
	ID            string     `json:"id"`
	TenantID      string     `json:"tenant_id"`
	TargetID      string     `json:"target_id"`
	TargetName    string     `json:"target_name"`
	TargetType    string     `json:"target_type"`
	Status        string     `json:"status"` // queued, running, completed, failed
	StartedAt     *time.Time `json:"started_at,omitempty"`
	CompletedAt   *time.Time `json:"completed_at,omitempty"`
	FindingsCount int        `json:"findings_count"`
	Error         string     `json:"error,omitempty"`
	ProgressPct   int        `json:"progress_pct"`
	CreatedAt     time.Time  `json:"created_at"`
}

// LeakFinding is a detected secret or credential exposure found during a scan.
type LeakFinding struct {
	ID             string     `json:"id"`
	TenantID       string     `json:"tenant_id"`
	JobID          string     `json:"job_id"`
	TargetID       string     `json:"target_id"`
	TargetName     string     `json:"target_name"`
	Severity       string     `json:"severity"` // critical, high, medium, low
	Type           string     `json:"type"`
	Description    string     `json:"description"`
	Location       string     `json:"location"`
	ContextPreview string     `json:"context_preview"`
	Entropy        float64    `json:"entropy"`
	// SecretFingerprint is the SHA-256 of the matched secret; the join key
	// the unified console uses to correlate a leaked credential with the KMS
	// key that protects it (via keycore's credential-binding registry).
	SecretFingerprint string  `json:"secret_fingerprint,omitempty"`
	Status         string     `json:"status"` // open, resolved, suppressed, false_positive
	DetectedAt     time.Time  `json:"detected_at"`
	ResolvedAt     *time.Time `json:"resolved_at,omitempty"`
	ResolvedBy     string     `json:"resolved_by,omitempty"`
	Notes          string     `json:"notes,omitempty"`
}

// CreateLeakTargetRequest is the request body for registering a new scan target.
type CreateLeakTargetRequest struct {
	Name    string `json:"name"`
	Type    string `json:"type"`
	URI     string `json:"uri"`
	Enabled *bool  `json:"enabled"`
}

// UpdateLeakFindingRequest is the request body for updating a finding's status/notes.
type UpdateLeakFindingRequest struct {
	Status     *string `json:"status"`
	ResolvedBy *string `json:"resolved_by"`
	Notes      *string `json:"notes"`
}
